File: Topcoder/Misc/Dynamic Programming/TCCC 2004 Online Round 1 - Flower Garden/src/FlowerGarden.java
import java.util.Arrays;

/*
 * author: sbaldrich
 * Source: TCCC2003 Semifinals Div1 - 1 - FlowerGarden
 * 
 * DP. At first one might think a simple sort routine could help, but since there is no way
 * to sort the flowers only by comparing pairs, dp is the way to go. At each iteration get 
 * the next possible flower, in order to do so, check each of the unprocessed flowers to see
 * if there is another flower that makes it impossible to place it on front. If there are various
 * flowers that could be placed on front, take the tallest. O(n^3) 
 */
public class FlowerGarden {

	public int[] getOrdering(int[] h, int[] b, int[] w) {

		int n = h.length, i, j, k;
		int ans[] = new int[n];
		boolean done[] = new boolean[n];
		Arrays.fill(done, false);

		for (i = 0; i < n; i++) {
			int best = -1;
			for (j = 0; j < n; j++) 
				if (!done[j]) {
					boolean ok = true;
					for (k = 0; k < n; k++) {
						if (k != j && !done[k] && h[k] < h[j] && b[k] <= w[j]
								&& w[k] >= b[j])
							ok = false;
					}
					if (ok && (best == -1 || h[j] > h[best]))
						best = j;
				}
			ans[i] = h[best];
			done[best] = true;
		}
		return ans;
	}
}

File: topcoder/tccc04/FlowerGarden.cpp
#include<bits/stdc++.h>

using namespace std;

struct flower{
  flower(int height, int bloom, int wilt) : height(height), bloom(bloom), wilt(wilt) {}
  int height, bloom, wilt;
  bool operator < (const flower& that) const{
    return this -> height < that.height;
  }
};

bool overlap(const flower& first, const flower& second){
  if (first.bloom <= second.bloom)
  return first.wilt >= second.bloom;
  return second.wilt >= first.bloom;
}

class FlowerGarden {
public:
vector <int> getOrdering(vector <int> height, vector <int> bloom, vector <int> wilt) {
  int n = height.size();
  vector<flower> flowers;
  vector<bool> planted = vector<bool>(n, false);
  vector<int> ans;
  int flowers_left = n;
  for(int i = 0; i < n; i++)
      flowers.push_back(flower(height[i], bloom[i], wilt[i]));
  sort(flowers.begin(), flowers.end());
  while(flowers_left--){
    int next = -1;
    for(int i = n - 1; i >= 0; i--){
      if(planted[i])
        continue;
      next = i;
      for(int j = 0; next != -1 && j < n; j++){
        if(i == j || planted[j])
          continue;
        if(overlap(flowers[i], flowers[j]) && flowers[i].height > flowers[j].height)
          next =- 1;
      }
      if(next != -1)
        break;
    }
    ans.push_back(flowers[next].height);
    planted[next] = true;
  }
  return ans;
  }
};
